source/spirv_stats.cpp

```diff
@@ -149,79 +130,74 @@ class StatsAggregator {
   }
 
   // Collects statistics of literal strings used by opcodes.
-  void ProcessLiteralStrings() {
-    const val::Instruction& inst = GetCurrentInstruction();
-    for (const auto& operand : inst.operands()) {
+  void ProcessLiteralStrings(const val::Instruction* inst) {
+    for (const auto& operand : inst->operands()) {
       if (operand.type == SPV_OPERAND_TYPE_LITERAL_STRING) {
         const std::string str =
-            reinterpret_cast<const char*>(&inst.words()[operand.offset]);
-        ++stats_->literal_strings_hist[inst.opcode()][str];
+            reinterpret_cast<const char*>(&inst->words()[operand.offset]);
+        ++stats_->literal_strings_hist[inst->opcode()][str];
       }
     }
   }
 
   // Collects statistics of all single word non-id operand slots.
-  void ProcessNonIdWords() {
-    const val::Instruction& inst = GetCurrentInstruction();
+  void ProcessNonIdWords(const val::Instruction* inst) {
     uint32_t index = 0;
-    for (const auto& operand : inst.operands()) {
+    for (const auto& operand : inst->operands()) {
       if (operand.num_words == 1 && !spvIsIdType(operand.type)) {
         ++stats_->operand_slot_non_id_words_hist[std::pair<uint32_t, uint32_t>(
-            inst.opcode(), index)][inst.word(operand.offset)];
+            inst->opcode(), index)][inst->word(operand.offset)];
       }
       ++index;
     }
   }
 
   // Collects OpCapability statistics.
-  void ProcessCapability() {
-    const val::Instruction& inst = GetCurrentInstruction();
-    if (inst.opcode() != SpvOpCapability) return;
-    const uint32_t capability = inst.word(inst.operands()[0].offset);
+  void ProcessCapability(const val::Instruction* inst) {
+    if (inst->opcode() != SpvOpCapability) return;
+    const uint32_t capability = inst->word(inst->operands()[0].offset);
     ++stats_->capability_hist[capability];
   }
 
   // Collects OpExtension statistics.
-  void ProcessExtension() {
-    const val::Instruction& inst = GetCurrentInstruction();
-    if (inst.opcode() != SpvOpExtension) return;
-    const std::string extension = GetExtensionString(&inst.c_inst());
+  void ProcessExtension(const val::Instruction* inst) {
+    if (inst->opcode() != SpvOpExtension) return;
+    const std::string extension = GetExtensionString(&inst->c_inst());
     ++stats_->extension_hist[extension];
   }
 
   // Collects OpCode statistics.
-  void ProcessOpcode() {
-    auto inst_it = vstate_->ordered_instructions().rbegin();
-    const SpvOp opcode = inst_it->opcode();
+  void ProcessOpcode(const val::Instruction* inst, size_t idx) {
```

Comment:
Please give this method a good look as it was doing things with iterators that I converted to indexes ...